parsons/catalist/catalist.py

```diff
@@ -0,0 +1,388 @@
+import logging
+import os
+import tempfile
+import time
+import urllib
+from typing import Optional, Union, Dict, List
+from zipfile import ZipFile
+
+import requests
+from parsons.etl import Table
+from parsons.sftp import SFTP
+from parsons.utilities.api_connector import APIConnector
+
+logger = logging.getLogger(__name__)
+
+# Currently the default and only available template for working with the Match API
+template_id = 48827
+
+
+class CatalistMatch:
+    """Connector for working with the Catalist Match API.
+
+    The Catalist Match tool requires OAuth2.0 client credentials for the API as well as
+    credentials for accessing the Catalist sftp bucket. Each Catalist client is given
+    their own bucket alias named after a tree species, used for constructing the
+    filepath within the sftp bucket.
+
+    Accessing the Catalist sftp bucket and Match API both require the source IP address
+    to be explicitly white-listed by Catalist.
+
+    Example usage:
+    ```
+    tbl = Table.from_csv(...)
+    client = CatalistMatch(...)
+    match_result = client.match(tbl)
+    ```
+
+    Note that matching can take from 10 minutes up to 6 hours or longer to complete, so
+    you may want to think strategically about how to await completion without straining
+    your compute resources on idling.
+
+    To separate submitting the job and fetching the result:
+    ```
+    tbl = Table.from_csv(...)
+    client = CatalistMatch(...)
+    response = client.upload(tbl)
+    match_result = client.await_completion(response["id"])
+    ```
+
+    """
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        sftp_username: str,
+        sftp_password: str,
+    ) -> None:
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.fetch_token()
+        self.connection = APIConnector("http://api.catalist.us/mapi/")
+        self.sftp = SFTP("t.catalist.us", sftp_username, sftp_password)
+
+    @property
+    def token(self) -> str:
+        """If token is not yet fetched or has expired, fetch new token."""
+        if not (self._token and time.time() < self._token_expired_at):
+            self.fetch_token()
+        return self._token
+
+    def fetch_token(self) -> None:
+        """Fetch auth0 token to be used with Catalist API."""
+        url = "https://auth.catalist.us/oauth/token"
+        payload = {
+            "grant_type": "client_credentials",
+            "audience": "catalist_api_m_prod",
+        }
+        response = requests.post(
+            url, json=payload, auth=(self.client_id, self.client_secret)
+        )
+        data = response.json()
+
+        self._token = data["access_token"]
+        self._token_expired_at = time.time() + data["expires_in"]
+
+        logger.info("Token refreshed.")
+
+    def load_table_to_sftp(
+        self, table: Table, input_subfolder: Optional[str] = None
+    ) -> str:
+        """Load table to Catalist sftp bucket.
```

Comment:
Can we get `Args:` in this docstring?

Reply:
Done with 561deccce2f6a219c0270c382c2906805ecdbd75